spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -285,4 +300,87 @@ protected void set(InternalRow struct, int pos, Object value) {
       }
     }
   }
+
+  private static class ComplexUnionReader implements ValueReader<InternalRow> {
```

Comment:
What is specific to Spark about this? Can we use an approach like the struct reader and have a generic one that is extended by Spark, Flink, etc. to make the type concrete?